packages/snaps-rpc-methods/src/preinstalled/middleware.ts

```diff
@@ -0,0 +1,80 @@
+import type { JsonRpcMiddleware } from '@metamask/json-rpc-engine';
+import type {
+  PermissionConstraint,
+  RequestedPermissions,
+} from '@metamask/permission-controller';
+import { isEqual } from '@metamask/snaps-utils';
+import { hasProperty, type Json, type JsonRpcParams } from '@metamask/utils';
+
+import { SnapEndowments } from '../endowments';
+
+export type PreinstalledSnapsMiddlewareHooks = {
+  getPermittedEvmAccounts: () => string[];
+  getAllEvmAccounts: () => string[];
+  getPermissions: () => Record<string, PermissionConstraint> | undefined;
+  grantPermissions: (permissions: RequestedPermissions) => void;
+};
+
+/**
+ * Creates a middleware that automatically grants account permissions to preinstalled Snaps
+ * that want to use the Ethereum provider endowment.
+ *
+ * @param hooks - The hooks used by the middleware.
+ * @param hooks.getPermittedEvmAccounts - Hook for retrieving the currently permitted EVM addresses.
+ * @param hooks.getAllEvmAccounts - Hook for retriveing all available EVM addresses.
+ * @param hooks.getPermissions - Hook for retrieving the permissions of the requesting origin.
+ * @param hooks.grantPermissions - Hook for granting permissions to the requesting origin.
+ * @returns The middleware.
+ */
+export function createPreinstalledSnapsMiddleware({
+  getPermittedEvmAccounts,
+  getAllEvmAccounts,
+  getPermissions,
+  grantPermissions,
+}: PreinstalledSnapsMiddlewareHooks): JsonRpcMiddleware<JsonRpcParams, Json> {
+  return function methodMiddleware(request, _response, next, _end) {
+    if (request.method.startsWith('snap')) {
+      return next();
+    }
+
+    const permissions = getPermissions();
+
+    if (
+      !permissions ||
+      !hasProperty(permissions, SnapEndowments.EthereumProvider)
+    ) {
+      return next();
+    }
+
+    const evmAccounts = getAllEvmAccounts();
+    const existingEvmAccounts = getPermittedEvmAccounts();
+
+    if (isEqual(evmAccounts, existingEvmAccounts)) {
+      return next();
+    }
+
+    grantPermissions({
+      'endowment:caip25': {
```

Comment:
Should this be merged with the existing permission (if any)?

Reply:
I mean, the existing value of the `authorizedScopes` value of the `endowment:caip25` permission (if any).

😓

Reply:
Done